### PR TITLE
[FLINK-35363] FLIP-449: Reorganization of flink-connector-jdbc

### PR DESCRIPTION
**Pull request by eskabetxe**
This is first phase of FLIP
each task has its own commit

tasks done:
https://issues.apache.org/jira/browse/FLINK-35370
https://issues.apache.org/jira/browse/FLINK-35364
https://issues.apache.org/jira/browse/FLINK-35365
https://issues.apache.org/jira/browse/FLINK-35366
https://issues.apache.org/jira/browse/FLINK-35367
https://issues.apache.org/jira/browse/FLINK-35494

### CODE REVIEW

File: flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/core/table/JdbcFactory.java
Comment (RocMarshal) [inline flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/core/table/JdbcFactory.java:37]: @eskabetxe Great work! 

Just left a few minor comments.
1 Maybe The original name is better for the followed items:
- Naming
- package structure
- The semantic of the class


2 The CI is failed , which is caused by something unknown possibly

Comment (eskabetxe) [reply flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/core/table/JdbcFactory.java:37]: point 1: I will change it as we talk offline..
point 2: Fixed, was a dependency problem//

Comment (eskabetxe) [reply flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/core/table/JdbcFactory.java]: done

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/AbstractDialect.java
Comment (RocMarshal) [inline flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/AbstractDialect.java:32]:
```suggestion
 * @deprecated Use {@link org.apache.flink.connector.jdbc.core.database.dialect.AbstractDialect}.
```
How about using the 'link' tag to point the target new Class ?

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/AbstractJdbcCatalog.java
Comment (RocMarshal) [inline flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/AbstractJdbcCatalog.java:24]:
```suggestion
 * @deprecated Use {@link org.apache.flink.connector.jdbc.core.database.catalog.AbstractJdbcCatalog}.
```

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/converter/AbstractJdbcRowConverter.java
Comment (RocMarshal) [inline flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/converter/AbstractJdbcRowConverter.java:27]:
```suggestion
 * @deprecated Use {@link AbstractDialectConverter}.
```

File: flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/internal/options/InternalJdbcConnectionOptions.java
Comment (RocMarshal) [inline flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/internal/options/InternalJdbcConnectionOptions.java:182]:
```suggestion
         * JdbcFactoryLoader#load(String, ClassLoader)} from DB url.
```

File: flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/core/datastream/Jdbc.java
Comment (RocMarshal) [inline flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/core/datastream/Jdbc.java:26]: The repo is lacking of documentation pages to describe the new jdbc sink2 and the current usage for here.
Would you like to supplement it by https://issues.apache.org/jira/browse/FLINK-35811 ?

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/JdbcCatalog.java
Comment (RocMarshal) [inline flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/JdbcCatalog.java:40]:
```suggestion
 * @deprecated Use {@link org.apache.flink.connector.jdbc.core.database.catalog.JdbcCatalog}.
```

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/factory/JdbcCatalogFactory.java
Comment (RocMarshal) [inline flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/factory/JdbcCatalogFactory.java:45]:
```suggestion
 * @deprecated Use {@link org.apache.flink.connector.jdbc.core.database.catalog.factory.JdbcCatalogFactory}.
```

File: flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/core/database/catalog/JdbcCatalogTypeMapper.java
Comment (RocMarshal) [inline flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/core/database/catalog/JdbcCatalogTypeMapper.java:30]: How about rename to 'JdbcTypeMapper' ?

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcConnectorOptions.java
Comment (RocMarshal) [inline flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcConnectorOptions.java:26]:
```suggestion
 * @deprecated Please use {@link org.apache.flink.connector.jdbc.core.table.JdbcConnectorOptions}.
```

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java
Comment (RocMarshal) [inline flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java:29]:
```suggestion
 * @deprecated Use {@link org.apache.flink.connector.jdbc.core.database.dialect.JdbcDialect}.
```

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialectFactory.java
Comment (RocMarshal) [inline flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialectFactory.java:33]:
```suggestion
 * @deprecated Use {@link JdbcFactory}.
```

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialectTypeMapper.java
Comment (RocMarshal) [inline flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialectTypeMapper.java:26]:
```suggestion
 * @deprecated Use {@link JdbcCatalogTypeMapper}.
```

Maybe it need the renamed name mentioned above~

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/converter/JdbcRowConverter.java
Comment (RocMarshal) [inline flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/converter/JdbcRowConverter.java:29]:
```suggestion
 * @deprecated Use {@link JdbcDialectConverter}.
```

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/source/JdbcSource.java
Comment (RocMarshal) [inline flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/source/JdbcSource.java:56]:
```suggestion
 * @deprecated Please use {@link org.apache.flink.connector.jdbc.core.datastream.source.JdbcSource}.
```

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/source/JdbcSourceBuilder.java
Comment (RocMarshal) [inline flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/source/JdbcSourceBuilder.java:93]:
```suggestion
 * @deprecated Please use {@link
 *     org.apache.flink.connector.jdbc.core.datastream.source.JdbcSourceBuilder}.
```

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/source/reader/extractor/ResultExtractor.java
Comment (RocMarshal) [inline flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/source/reader/extractor/ResultExtractor.java:29]:
```suggestion
 * @deprecated Please use {@link
```

**Review by RocMarshal — COMMENTED** [review]
Thanks @eskabetxe for the hard work!
Looks nice to me on the functional refactor.
Just left a few of comments about doc/naming.
Please let me know what's your opinion :)

File: flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/core/datastream/Jdbc.java
Comment (eskabetxe) [reply flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/core/datastream/Jdbc.java:26]: I think that all doc should also be refactored, maybe have one sub-page per connector, and clean all deprecated docs.
We could start talk about it, but it is on the plan.

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/AbstractDialect.java
Comment (eskabetxe) [reply flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/AbstractDialect.java:32]: done

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/converter/AbstractJdbcRowConverter.java
Comment (eskabetxe) [reply flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/converter/AbstractJdbcRowConverter.java:27]: done

File: flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/internal/options/InternalJdbcConnectionOptions.java
Comment (eskabetxe) [reply flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/internal/options/InternalJdbcConnectionOptions.java:182]: changed to JdbcFactoryLoader#loadDialect

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/JdbcCatalog.java
Comment (eskabetxe) [reply flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/JdbcCatalog.java:40]: done

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/factory/JdbcCatalogFactory.java
Comment (eskabetxe) [reply flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/factory/JdbcCatalogFactory.java:45]: done

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcConnectorOptions.java
Comment (eskabetxe) [reply flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcConnectorOptions.java:26]: done

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java
Comment (eskabetxe) [reply flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java:29]: done

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialectFactory.java
Comment (eskabetxe) [reply flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialectFactory.java:33]: done

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialectTypeMapper.java
Comment (eskabetxe) [reply flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialectTypeMapper.java:26]: done

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/converter/JdbcRowConverter.java
Comment (eskabetxe) [reply flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/converter/JdbcRowConverter.java:29]: done

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/source/JdbcSource.java
Comment (eskabetxe) [reply flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/source/JdbcSource.java:56]: fixed

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/source/JdbcSourceBuilder.java
Comment (eskabetxe) [reply flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/source/JdbcSourceBuilder.java:93]: fixes

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/source/reader/extractor/ResultExtractor.java
Comment (eskabetxe) [reply flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/source/reader/extractor/ResultExtractor.java:29]: fixed

File: flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/core/database/catalog/JdbcCatalogTypeMapper.java
Comment (eskabetxe) [reply flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/core/database/catalog/JdbcCatalogTypeMapper.java:30]: we use the same pattern on all Jdbc{kink}Anything, where kind is source, sink, dialect, catalog (already from before the refactor)

this was called "JdbcDialectTypeMapper"

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/AbstractJdbcCatalog.java
Comment (eskabetxe) [reply flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/AbstractJdbcCatalog.java]: done

**Comment by eskabetxe** [comment]
@leonardBang @1996fanrui @snuyanzin can you check please when you have time

70% of the refactor has been completed.
The deprecation of the old sources and sinks is still pending, as they are still used in the table API. 
The idea is to review and merge this part (there are already significant changes) and continue working on other PRs to:

- Modify the table API for the new sources and sinks.
- Deprecate and move the old sources and sinks.
- Finish organizing the code in the core module.
- Reorganize the documentation.

Note: Before the merge, we can consider removing the backward-compatibility module to avoid generating the jar, as it was only meant to be a helper.

**Comment by RocMarshal** [comment]
> Note: Before the merge, we can consider removing the backward-compatibility module (and workflow) to avoid generating the jar, as it was only meant to be a helper.

Hi, @eskabetxe 
Considering that this is a relatively large refactoring action,
Perhaps retaining this module can also help developers quickly verify and fix compatibility related issues.
How about continuing to keep one or two versions of the cycle? In other words, keep this module until the refactored code becomes stable and mature.
Please let me know what's your opinion.

**Review by RocMarshal — APPROVED** [review]
Thanks @eskabetxe for the hard work.
LGTM +1. 
now ping @1996fanrui

**Comment by 1996fanrui** [comment]
Thanks @eskabetxe for the contribution, and thanks @RocMarshal for the review!

Due to both of you are active contributor of [flink-connector-jdbc](https://github.com/apache/flink-connector-jdbc), and you aren't committer. So I will merge this PR if no objection before next Monday(in 5 days).

**Pull request merged**
